### ISSUE TITLE
Blame is very slow

### ISSUE THREAD

**Issue by mkhadilk** [go-git/go-git#14]
I tested a Blame and as compared to CLI it is very slow. CLI takes around < 1 and go-git takes 35 second.

**Comment by mkhadilk** [comment]
I would have helped by fork and an implementation by diff. But, may not get time till May. And why looking at the code, I realized that FilePatches does not give me proper 'rename'.

I was expecting that "from, to" in the case of rename will be correct. But it is not. That is critical as blame depends on that.

**Comment by mkhadilk** [comment]
type FilePatch interface {
	// IsBinary returns true if this patch is representing a binary file.
	IsBinary() bool
	// Files returns the from and to Files, with all the necessary metadata to
	// about them. If the patch creates a new file, "from" will be nil.
	// If the patch deletes a file, "to" will be nil.
	Files() (from, to File)
	**// Chunks returns a slice of ordered changes to transform "from" File to
	// "to" File. If the file is a binary one, Chunks will be empty.**
	Chunks() []Chunk
}

But what is it is a rename?

**Comment by mcuadros** [comment]
I am not the original author of this part, and It's a long time since I don't touch this part. But what I remember is that the rename it's no handled because it's not part of the diff process it's a heuristic with a threshold, after the process of making the diff.

**Comment by erizocosmico** [comment]
@mkhadilk rename detection has just landed in #38

**Comment by patrickdevivo** [comment]
I noticed this as well and ended up shelling out to `git blame ...` as a more performant alternative. The code for which can be found here: https://github.com/augmentable-dev/tickgit/blob/master/pkg/blame/blame.go (probably worth adding a README to that), but just leaving a note in case anyone gets blocked/stuck on the performance of blame in go-git!

**Comment by mkhadilk** [comment]
I did exactly the same by doing exec but then that also becomes a problem as need to parse the output...

**Comment by cyw3** [comment]
Is there any way to speed up?

**Comment by donatj** [comment]
Not to pile on, but I work on a pretty large repo. Using Blame here on our README.md takes over 20 seconds. I really thought I was doing something wrong until I found this ticket. 

Compare `git blame` to this

![image](https://user-images.githubusercontent.com/133747/184722044-6cda9066-8363-4c1d-a332-9c89127f2d43.png)

![image](https://user-images.githubusercontent.com/133747/184722179-52d487d7-978c-4e2b-bcf2-1bef39141b47.png)

**Comment by pjbgf** [comment]
The blame feature was rewrote (#789) since the last activity on this issue and introduced some [performance improvements](https://github.com/go-git/go-git/pull/789#issuecomment-1622616280). So I am closing it for the time being. If the issue persists we can either reopen this or create a new one.

**Comment by pjbgf** [comment]
@iainjreid can you please confirm what version of go-git you are using?

**Comment by iainjreid** [comment]
@pjbgf `v5.12.0`

```
// go.sum
github.com/go-git/go-git/v5 v5.12.0 h1:7Md+ndsjrzZxbddRDZjF14qK+NN56sy6wkqaVrjZtys=
github.com/go-git/go-git/v5 v5.12.0/go.mod h1:FTM9VKtnI2m65hNI/TenDDDnUf2Q9FHnXYjuz9i5OEY=
```

**Comment by iainjreid** [comment]
After some more testing using the in-memory storage (which is immeasurably faster, but the bottleneck still persists), I believe this issue is simply made apparent and worse by my SQL-backed storage.

My gut feeling is that as the commit history is being walked, the round trip cost of whichever storage mechanism is being used becomes an overwhelmingly high overhead?

I'll run some tests locally and report back 👍

**Comment by donatj** [comment]
I am patiently waiting for @iainjreid to report back so you should keep this open

**Comment by pjbgf** [comment]
I can see there was a minor improvement in time (ns/op) from `v5.12` onwards:

![image](https://github.com/user-attachments/assets/f26c488f-9ed4-467e-a0f3-9f747316c477)

Although the allocations per op decreased by roughly 100k.
![image](https://github.com/user-attachments/assets/6668cb3b-ee84-4f6a-89e4-32d0ff74a83c)

Marking this as no-autoclose so we can investigate this forward. However, any changes would target v6.

**Comment by martinlehoux** [comment]
Hi, I'd be happy to help if I can. As others, I implemented an alternative version using the git cli, but of course it would make more sense to improve go-git implementation.
Maybe having a look at git implementation could help us understand what they do that is better

**Comment by pjbgf** [comment]
@martinlehoux feel free to propose a PR and tag me for a review. I made a few [changes](https://github.com/go-git/go-git/pull/1523) on the packfile delta processing, however that does not seem to have an impact on this feature.

**Comment by pjbgf** [comment]
This [PR](https://github.com/go-git/go-git/pull/1523) (WIP) focuses in improving memory churn around the packfile and storage, and in my tests it seems to have a very positive impact on Blame. Running against this repository it is roughly 50% faster. It would be nice if other folks could confirm the improvement on their use cases.

**Comment by runxiyu** [comment]
Blame could also benefit from commit-graph's bloom filters.

**Comment by pjbgf** [comment]
@runxiyu that's right. AFAIK, that's a feature we are yet to implement in `plumbing/format/commitgraph`.

**Comment by runxiyu** [comment]
I should be able to port one right after my exams

**Comment by onee-only** [comment]
> Blame could also benefit from commit-graph's bloom filters.

Related: https://github.com/go-git/go-git/issues/144
